File: langchain-agent/src/agent.ts
// Copyright (c) Microsoft Corporation.
// Licensed under the MIT License.

import { TurnState, Authorization, AgentApplication, TurnContext, MemoryStorage } from '@microsoft/agents-hosting';
import { ActivityTypes } from '@microsoft/agents-activity';
import { getObservabilityAuthenticationScope } from '@microsoft/agents-a365-runtime';
import tokenCache, { createAgenticTokenCacheKey } from './token-cache.js';
import { Client, getClient } from './client.js';
import { AgenticTokenCacheInstance, BaggageBuilderUtils, A365_PARENT_SPAN_KEY } from '@microsoft/agents-a365-observability-hosting';
import {
  BaggageBuilder,
  InvokeAgentScope,
  InvokeAgentScopeDetails,
  AgentDetails,
  Request as A365Request,
  ParentSpanRef,
} from '@microsoft/agents-a365-observability';

export class A365Agent extends AgentApplication<TurnState> {
  static authHandlerName: string = 'agentic';

  constructor() {
    const useAgenticAuth = process.env.USE_AGENTIC_AUTH === 'true';

    super({
      startTypingTimer: true,
      storage: new MemoryStorage(),
      ...(useAgenticAuth && {
        authorization: {
          agentic: {
            type: 'agentic',
          } // scopes set in the .env file or environment
        }
      })
    });

    this.onActivity(ActivityTypes.Message, async (context: TurnContext, state: TurnState) => {
      await this.handleAgentMessageActivity(context, state);
    });
  }

  /**
   * Handles incoming user messages and sends responses.
   * Creates an InvokeAgentScope and stores its span context in turnState
   * so that OutputLoggingMiddleware links output spans as children.
   */
  async handleAgentMessageActivity(turnContext: TurnContext, _state: TurnState): Promise<void> {
    const userMessage = turnContext.activity.text?.trim() || '';
    if (!userMessage) {
      return;
    }

    // Create InvokeAgentScope to trace the full agent invocation
    const request: A365Request = {
      conversationId: turnContext.activity.conversation?.id,
    };
    const invokeScopeDetails: InvokeAgentScopeDetails = {};
    const agentDetails: AgentDetails = {
      agentId: turnContext.activity.recipient?.agenticAppId || 'langchain-agent',
      agentName: 'LangChainA365Agent',
      tenantId: turnContext.activity.recipient?.tenantId || 'unknown',
    };

    const invokeScope = InvokeAgentScope.start(request, invokeScopeDetails, agentDetails);
    try {
      // Preload observability token so the exporter can resolve agent identity
      await this.preloadObservabilityToken(turnContext);

      // Store the span context so OutputLoggingMiddleware links output spans as children
      const spanCtx = invokeScope.getSpanContext();
      const parentSpanRef: ParentSpanRef = {
        traceId: spanCtx.traceId,
        spanId: spanCtx.spanId,
        traceFlags: spanCtx.traceFlags,
      };
      turnContext.turnState.set(A365_PARENT_SPAN_KEY, parentSpanRef);

      const client: Client = await getClient();
      const response = await client.invokeAgent(userMessage);

      // Send the response back to the user
      await turnContext.sendActivity(response);
    } catch (error) {
      invokeScope.recordError(
        error instanceof Error ? error : new Error(String(error))
      );
      console.error('Agent invocation error:', error);
      await turnContext.sendActivity(`Error: ${error instanceof Error ? error.message : String(error)}`);
    } finally {
      invokeScope.dispose();
    }
  }

  /**
   * Preloads or refreshes the Observability token used by the Agent 365 Observability exporter.
   */
  private async preloadObservabilityToken(turnContext: TurnContext): Promise<void> {
    const authorization = this.getAuthorizationSafe();

    // Only attempt token preloading if authorization is configured
    if (!authorization) {
      console.log('[A365Agent] Authorization not configured, skipping observability token preload');
      return;
    }

    const agentId = turnContext?.activity?.recipient?.agenticAppId ?? '';
    const tenantId = turnContext?.activity?.recipient?.tenantId ?? '';

    if (process.env.Use_Custom_Resolver === 'true') {
      const aauToken = await authorization.exchangeToken(turnContext, 'agentic', {
        scopes: getObservabilityAuthenticationScope()
      });
      console.log(`Preloaded Observability token for agentId=${agentId}, tenantId=${tenantId} token=${aauToken?.token?.substring(0, 10)}...`);
      const cacheKey = createAgenticTokenCacheKey(agentId, tenantId);
      tokenCache.set(cacheKey, aauToken?.token || '');
    } else {
      await AgenticTokenCacheInstance.RefreshObservabilityToken(
        agentId,
        tenantId,
        turnContext,
        authorization,
        getObservabilityAuthenticationScope()
      );
    }
  }

  private getAuthorizationSafe(): Authorization | undefined {
    try {
      return this.authorization as Authorization;
    } catch {
      return undefined;
    }
  }
}

export const agentApplication = new A365Agent();

File: python/distro/langchainsample/src/agent.py
# Copyright (c) Microsoft Corporation. All rights reserved.
# Licensed under the MIT License.

import logging
from os import environ

from dotenv import load_dotenv
from microsoft_agents.activity import load_configuration_from_env
from microsoft_agents.authentication.msal import MsalConnectionManager
from microsoft_agents.hosting.aiohttp import CloudAdapter
from microsoft_agents.hosting.core import (
    AgentApplication,
    Authorization,
    MemoryStorage,
    TurnContext,
    TurnState,
)
from microsoft_agents.hosting.core.storage import (
    ConsoleTranscriptLogger,
    TranscriptLoggerMiddleware,
)
from microsoft.opentelemetry.a365.core import (
    AgentDetails,
    BaggageBuilder,
    CallerDetails,
    Channel,
    ChatMessage,
    InvokeAgentScope,
    InvokeAgentScopeDetails,
    MessageRole,
    OutputMessage,
    OutputMessages,
    OutputScope,
    Request,
    Response,
    ServiceEndpoint,
    SpanDetails,
    TextPart,
    UserDetails,
)
from microsoft.opentelemetry.a365.hosting.token_cache_helpers import AgenticTokenStruct
from microsoft_agents_a365.runtime.environment_utils import (
    get_observability_authentication_scope,
)
from services.langchain_service import call_langchain
from utils.observability_helpers import (
    create_agent_details,
    create_request_details,
)
from utils.token_cache import cache_agentic_token

logger = logging.getLogger(__name__)

load_dotenv()
agents_sdk_config = load_configuration_from_env(environ)


STORAGE = MemoryStorage()
CONNECTION_MANAGER = MsalConnectionManager(**agents_sdk_config)
ADAPTER = CloudAdapter(connection_manager=CONNECTION_MANAGER)
ADAPTER.use(TranscriptLoggerMiddleware(ConsoleTranscriptLogger()))
AUTHORIZATION = Authorization(STORAGE, CONNECTION_MANAGER, **agents_sdk_config)

AGENT_APP = AgentApplication[TurnState](
    storage=STORAGE, adapter=ADAPTER, authorization=AUTHORIZATION, **agents_sdk_config
)

# Token cache will be injected via adapter.app_context by start_server


def _get_token_cache():
    """Get token cache from application context (injected by start_server)."""
    if hasattr(ADAPTER, "app_context") and "token_cache" in ADAPTER.app_context:
        return ADAPTER.app_context["token_cache"]
    return None


@AGENT_APP.activity("message", auth_handlers=["AGENTIC"])
async def on_message(context: TurnContext, _state: TurnState):
    """Handle incoming messages and respond using LangChain with Azure OpenAI."""
    agent_details = create_agent_details(context)

    # Register observability token cache using injected AgenticTokenCache
    token_cache = _get_token_cache()
    if token_cache:
        token_struct = AgenticTokenStruct(
            authorization=AGENT_APP.auth,
            turn_context=context,
            auth_handler_name="AGENTIC",
        )

        token_cache.register_observability(
            agent_id=agent_details.agent_id,
            tenant_id=agent_details.tenant_id,
            token_generator=token_struct,
            observability_scopes=get_observability_authentication_scope(),
        )
    else:
        logger.warning("Token cache not available in app context")

    # Create invoke agent scope details
    invoke_scope_details = InvokeAgentScopeDetails()

    # Get the user's message early for request details
    user_message = context.activity.text
    if not user_message:
        await context.send_activity("I didn't receive any message text to process.")
        return

    # Create request details
    request_details = create_request_details(
        user_message, context.activity.conversation.id if context.activity.conversation else None
    )

    try:
        # Start invoke agent scope for observability
        invoke_scope = InvokeAgentScope.start(
            request=request_details,
            scope_details=invoke_scope_details,
            agent_details=agent_details,
        )

        try:
            with invoke_scope:
                # Get the agentic user token with observability scope
                invoke_scope.record_input_messages([user_message])
                exaau_token = await AGENT_APP.auth.exchange_token(
                    context,
                    scopes=get_observability_authentication_scope(),
                    auth_handler_id="AGENTIC",
                )

                # Cache the agentic token for exporter use (old approach - for reference)
                cache_agentic_token(
                    agent_details.tenant_id, agent_details.agent_id, exaau_token.token
                )

                logger.info(f"Processing user message: {user_message}")

                # Call LangChain service - auto-instrumented by LangChainInstrumentor
                ai_response = await call_langchain(user_message, context)

                # Send the AI response back to the user
                await context.send_activity(ai_response)

            # Record output in a separate OutputScope outside the invoke scope,
            # linked to the invoke scope via parent_id
            invoke_span = invoke_scope._span
            if invoke_span and invoke_span.context:
                parent_id = (
                    f"00-{invoke_span.context.trace_id:032x}-{invoke_span.context.span_id:016x}-01"
                )
            else:
                parent_id = None

        except Exception as e:
            logger.error(f"Error in on_message: {e}")
            invoke_scope.record_error(e)
            await context.send_activity(
                "I encountered an error while processing your message. Please try again."
            )
    except Exception as e:
        logger.error(f"Error setting up observability scope: {e}")


@AGENT_APP.error
async def on_error(context: TurnContext, error: Exception):
    logger.error("[on_turn_error] unhandled error: %s", error)

    # Send a message to the user
    await context.send_activity("The agent encountered an error.")
